### ISSUE TITLE
Restarted kubelet should not evict pods due to the node affinity

### ISSUE THREAD

**Issue by gjkim42** [kubernetes/kubernetes#124586]
### What happened?

Currently, a running pod is being evicted if the assigned node, which does not meet the pod's node affinity requirements anymore, restarts its kubelet.

xref: https://github.com/kubernetes/kubernetes/issues/123980#issuecomment-2071554998

### What did you expect to happen?

https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#node-affinity

![Screenshot 2024-04-23 at 3 51 20 PM](https://github.com/kubernetes/kubernetes/assets/19814047/5b15fcd0-3601-4b77-b863-6ca0970b107b)

According to the documentation, node affinity should only affect the scheduling phase. Therefore, the running pod should continue to run without interruption.

### How can we reproduce it (as minimally and precisely as possible)?

1. Create a deployment with node affinity.
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: test
  name: test
spec:
  replicas: 1
  selector:
    matchLabels:
      app: test
  template:
    metadata:
      labels:
        app: test
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: foo
                operator: In
                values:
                - bar
      containers:
      - image: nginx
        imagePullPolicy: Always
        name: nginx
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
```
2. `kubectl label node TARGET_NODE foo=bar --overwrite` so that the pod is scheduled in the TARGET_NODE.
3. Waiting for the pod to be scheduled and running.
4. `kubectl label node TARGET_NODE foo=nonbar --overwrite`
5. Restart the kubelet of the TARGET_NODE.

### Anything else we need to know?

previous discussion: https://github.com/kubernetes/kubernetes/pull/101218#discussion_r627495203

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
Client Version: v1.28.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.31.0-alpha.0.470+7d880fd489d1e5-dirty
```

</details>

### Cloud provider

<details>
kind cluster
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by gjkim42** [comment]
https://github.com/kubernetes/kubernetes/blob/7d880fd489d1e57bc1c13568f1cbb4c5c5e93947/staging/src/k8s.io/api/core/v1/types.go#L3458-L3464

^
The other document is saying that the pod may be evicted due to the node affinity.

**Comment by gjkim42** [comment]
We can fix the behavior or documents.

The fix seems simple but may break many existing use cases. We may need some discussion about this.

**Comment by ffromani** [comment]
/triage accepted

In hindsight, the labelling part (xref: https://github.com/kubernetes/kubernetes/issues/123980#issuecomment-2071554998) deserves its own issue.

**Comment by gjkim42** [comment]
/cc @liggitt @alculquicondor 

since we had a similar discussion before in https://github.com/kubernetes/kubernetes/pull/101218#discussion_r627495203

**Comment by rmohr** [comment]
> The verification is happening both in the scheduler and in kubelet, at startup, but not during execution.

> Yes, we can update the documentation to be more clear, but we can't change the behavior.

Do you mean you can not change the behaviour to support `IgnoreDuringExecution` as documented right now? If so, could you elaborate on that a little bit?

**Comment by alculquicondor** [comment]
We are ignoring during execution.
In the case reported, the pod is marked as Failed before execution.

In other words, we are honoring what the API says. But there is an intermediate stage that the API doesn't say anything about: the time between pod scheduled and before it starts executing.

That's what needs to be clarified in the documentation, and it should match the existing behavior that has been there since this feature first launched. We cannot change that behavior or it would be backwards incompatible.

**Comment by rmohr** [comment]
> In the case reported, the pod is marked as Failed before execution.

Ok, trying to rephrase with my own words to be sure I understand:
 1. kubelet stop
 2. pod gets scheduled (but not started yet)
 3. kubelet starts and sees that the pod should not be started here (and is not yet executing)
 4. kubelet stops it
 
 That sounds fair to keep.

**Comment by Homura222** [comment]
if pod use `preferredDuringSchedulingIgnoredDuringExecution` nodeAffinity not `requiredDuringSchedulingIgnoredDuringExecution ` nodeAffinity , pod will not be killed in this case. Is this expected?

**Comment by alculquicondor** [comment]
Yes

**Comment by alculquicondor** [comment]
@SergeyKanzhelev can you chime in the kubelet details?

**Comment by rmohr** [comment]
> If we have to document the behavior as it works now, requiredDuringSchedulingIgnoredDuringExecution affects the pod during the scheduling time, the pod start time, AND the kubelet restart time. (kubelet restart time? why?)

+1, if the kubelet restart evicts based on this, then it is not `IgnoredDuringExecution` from a workload perspective.

**Comment by haircommander** [comment]
SIG node discussed this in our CI meeting. It could be technically possible to keep the state of node affinity for pods between restarts (or query the runtime before reattempting admission) but both of these would be invasive and tricky. Our suggestion would be to update the documentation to be more clear about what the kubelet does. Some of the language is passive ("kubelet may or may not ...") and maybe we should be more clear about what it does.

**Comment by rmohr** [comment]
> SIG node discussed this in our CI meeting. It could be technically possible to keep the state of node affinity for pods between restarts (or query the runtime before reattempting admission) but both of these would be invasive and tricky. Our suggestion would be to update the documentation to be more clear about what the kubelet does. Some of the language is passive ("kubelet may or may not ...") and maybe we should be more clear about what it does.

Does this mean that you are saying that the `IgnoreDuringExecution` part will not be fixed in case that the kubelet crashes?

**Comment by haircommander** [comment]
I'm not personally privvy to why "IgnoreDuringExecution" part was added. Are there other cases that it could potentially be ignored or is kubelet restart the only case?

**Comment by SergeyKanzhelev** [comment]
Triage notes:

- We definitely need to update documentation to tell what is the behavior today.
- Since it is hard to tell if anybody may rely on this behavior (re-apply affinity labels validation by restarting kubelet) - fixing this behavior is likely fall under a feature development category, not a bug.

/kind documentation
/remove-king bug

If somebody up to help with this, please file a separate feature tracking issue and/or a KEP.

**Comment by HirazawaUi** [comment]
This issue seems suitable to be discussed together with the content in https://github.com/kubernetes/enhancements/pull/5493

Both are related to kubelet restarts, with the goal of ensuring that running workloads are not affected after a kubelet restart. We could expand the scope of https://github.com/kubernetes/enhancements/pull/5493 to include preventing pod eviction during kubelet restarts...

@SergeyKanzhelev WDYT?

-----
EDIT:
After carefully tracing the context of this issue, it is not the same category of problem as kubernetes/enhancements#5493, and we should not group them together for discussion.

**Comment by HirazawaUi** [comment]
> It could be technically possible to keep the state of node affinity for pods between restarts (or query the runtime before reattempting admission) but both of these would be invasive and tricky.

After reviewing the overall code, I found that this issue is not as tricky as it initially seemed. I decided to check the pod's status on the node during the node affinity admission process (if it has been running on this node before), PR: https://github.com/kubernetes/kubernetes/pull/133803 .

Moreover, protecting pods from being rejected during admission after kubelet restarts is not unprecedented here, as we have already implemented similar logic elsewhere.
ref: https://github.com/kubernetes/kubernetes/blob/3026020b44a097a6f95a884e1e7f966c12675c0f/pkg/kubelet/cm/devicemanager/manager.go#L584-L597



> If somebody up to help with this, please file a separate feature tracking issue and/or a KEP.

Considering that this fix is not particularly complex, I don't believe it requires a KEP. I think we can add a feature gate starting from the beta stage and document the purpose of this feature gate. Would that be sufficient to alert users? If users rely on this functionality (bug), they can set the feature gate to false after reading the documentation or release notes to maintain the previous logic and provide timely feedback on any issues.

**Comment by rmohr** [comment]
> /remove-kind bug

Could there be a rationale added why this is considered not to be a bug? Thanks!

**Comment by natasha41575** [comment]
> > /remove-kind bug
> 
> Could there be a rationale added why this is considered not to be a bug? Thanks!

This was in response to Sergey's above comment: https://github.com/kubernetes/kubernetes/issues/124586#issuecomment-2917107610. It looks like he'd just made a typo.

**Comment by SergeyKanzhelev** [comment]
> > /remove-kind bug
> 
> Could there be a rationale added why this is considered not to be a bug? Thanks!

This was the rationale: https://github.com/kubernetes/kubernetes/issues/124586#issuecomment-2096412461


Can you explain the scenario? Why node have new labels that are inconsistent withe the Pods running on it?

**Comment by rmohr** [comment]
> Can you explain the scenario? Why node have new labels that are inconsistent withe the Pods running on it?

Sure:

 * When in the future we want pods to go to other nodes but without disrupting the ones running there (examples would be long running AI jobs, or CI jobs where you want to shift some workloads but not disrupt current workloads)
 * When we want to not schedule specific pods temporarily to a node (workload specific requirements not met for "new" workloads). There are many cases where a node is in principle still ready, but some specific sub-part for specific workloads are temporarily not in a shape to receive new workloads, but current ones are fine. For instance in kubevirt we add labels based on our own internal checks for readiness for VMs (that's different to general workload readiness).

We basically make use of the `IgnoreDuringExecution` part (which has it even in the name that it should leave them alone). And it works, except if the kubelet crashes or restarts. If it would be named `RequiredDuringSchedulingRequiredDuringExecution` we would not have built on it.

What people see a lot is this:
 1. they redo some labels, or some process change some labels
 2. at some unrelated point in time later, suddenly a lot of pods get evicted for unclear reasons, leading to temporary reduced availability of services. This is a problem for meeting SLOs (would be better for us to reserve unintended downtime due to e.g. node failure)

We then later reconstuct from logs that at some point the kubelet got reconfigured or restarted due to unrelated things.

**Comment by gjkim42** [comment]
/unassign 
since I don't have the bandwidth for this.

**Comment by nkawli** [comment]
We’ve encountered a critical issue after integrating a KubeVirt solution: running VM workloads are being evicted whenever the kubelet restarts—whether the restart is planned or unplanned—which results in service disruption.

We never noticed this problem with our regular non-VM workloads because those pods were already bound to nodes, and our node labels remained mostly static. However, with KubeVirt, the virt-handler updates node labels dynamically (see https://github.com/kubevirt/kubevirt/issues/11843), and this appears to trigger the eviction of running workloads due to the behavior described in the referenced issue.

This eviction behavior is impacting our ability to proceed with a VM-based approach with KubeVirt, and are looking for guidance or potential workarounds to proceed.

**Comment by SergeyKanzhelev** [comment]
As discussed at SIG Node meeting today, this can be unblocked by:

1. List all pod admission behaviors we have a desire to supress for already running Pods.
2. Look into possible race condition when Pod is scheduled, but was not registered with the API server as "admitted". And whether failing admission of a Pod in this case is reasonable.
3. Look into security implications of omitting this check on kubelet side (fake admitted status skips the taint validation)
4. Refresh the connected PR

Note, there are related issues:

1. https://github.com/kubernetes/kubernetes/pull/139959 (https://github.com/kubernetes/kubernetes/issues/125579) - "guaranteed" readmission when Device Plugin is slow to restart
2. Similar with memory manager chackpoint: https://github.com/kubernetes/kubernetes/issues/123971, https://github.com/kubernetes/kubernetes/issues/138495
3. https://github.com/kubernetes/kubernetes/issues/125815 - pod may be re-admitted and evicted immediately

**Comment by lukaszwojciechowski** [comment]
@SergeyKanzhelev 
~~Please add one more scenario:
4. https://github.com/kubernetes/kubernetes/issues/142328 - pod scheduled on node supporting feature via NDF, but evicted during re-admission after kubelet restar with FG being disabled~~

[UPDATE] After discussion on slack, it was decided that above scenario won't be considered. NDF won't get enhanced with possibility for feature to avoid re-admission eviction

**Comment by HirazawaUi** [comment]
> As discussed at SIG Node meeting today, this can be unblocked by:
> 
> 1. List all pod admission behaviors we have a desire to supress for already running Pods.
> 2. Look into possible race condition when Pod is scheduled, but was not registered with the API server as "admitted". And whether failing admission of a Pod in this case is reasonable.
> 3. Look into security implications of omitting this check on kubelet side (fake admitted status skips the taint validation)
> 4. Refresh the connected PR
> 
> Note, there are related issues:
> 
> 1. [Deferred Pod Admission for Unregistered Device Plugins #139959](https://github.com/kubernetes/kubernetes/pull/139959) ([Node reboot leaving existing pod using resources stuck with error UnexpectedAdmissionError  #125579](https://github.com/kubernetes/kubernetes/issues/125579)) - "guaranteed" readmission when Device Plugin is slow to restart
> 2. Similar with memory manager chackpoint: [pod with initcontainer failed with UnexpectedAdmissionError when restart kubelet #123971](https://github.com/kubernetes/kubernetes/issues/123971), [Memory Manager static policy rejects TopologyManager best-effort cross-NUMA hints causing UnexpectedAdmissionError #138495](https://github.com/kubernetes/kubernetes/issues/138495)
> 3. [Pod starting order after restart doesn't follow pod priorities #125815](https://github.com/kubernetes/kubernetes/issues/125815) - pod may be re-admitted and evicted immediately

@SergeyKanzhelev Then I think we need to start this from a KEP, clearly documenting our implementation path, the discussion process, and the impact of making this change. This KEP won't be very complex. We can add a deprecated feature gate, but then we'd have to wait another three months to implement it...
